src/packets.toit

```diff
@@ -135,100 +228,140 @@ class ConnectPacket extends Packet:
+      packet_id = Packet.decode_uint16 reader
       size -= 2
     else:
       packet_id = null
     payload = reader.read_bytes size
-    super TYPE
+    super TYPE --flags=(qos << 1) | (retain ? 1 : 0) | (duplicate ? 0b1000 : 0)
 
-  constructor .topic .payload --.qos/int --retain --.packet_id:
+  constructor .topic .payload --qos/int --retain/bool --.packet_id --duplicate=false:
     super TYPE
-      --flags=(qos << 1) | (retain ? 1 : 0)
+        --flags=(qos << 1) | (retain ? 1 : 0) | (duplicate ? 0b1000 : 0)
 
   variable_header -> ByteArray:
     buffer := bytes.Buffer
     Packet.encode_string buffer topic
-    if packet_id:
-      data := ByteArray 2
-      binary.BIG_ENDIAN.put_uint16 data 0 packet_id
-      buffer.write data
+    if packet_id: buffer.write (Packet.encode_uint16 packet_id)
     return buffer.bytes
 
-class PubAckPacket extends Packet implements PacketIDAck:
+  retain -> bool: return flags & 0b1 != 0
+  qos -> int: return (flags >> 1) & 0b11
+  duplicate -> bool: return flags & 0b1000 != 0
+
+  /**
+  Creates a new packet with the provided parameters replacing the current values.
+
+  Optional arguments don't work if one wants to pass in null. As such, the
+    $packet_id must be given as integer. If the packet-id should be null, then '-1' should
+    be used instead.
+  */
+  with -> PublishPacket
+      --topic/string?=null
+      --payload/ByteArray?=null
+      --qos/int?=null
+      --retain/bool?=null
+      --packet_id/int=-2
+      --duplicate/bool?=null:
+    new_packet_id /int? := ?
+    if packet_id == -2: new_packet_id = this.packet_id
+    else if packet_id == -1: new_packet_id = null
+    else: new_packet_id = packet_id
+    return PublishPacket
+        topic or this.topic
+        payload or this.payload
+        --qos = qos or this.qos
+        --retain = retain != null ? retain : this.retain
+        --packet_id = new_packet_id
+        --duplicate = duplicate != null ? duplicate : this.duplicate
+
+class PubAckPacket extends Packet:
   static TYPE ::= 4
 
   packet_id/int
 
-  constructor .packet_id:
+  constructor --.packet_id:
     super TYPE
 
-  constructor.deserialize reader/reader.BufferedReader:
-    data := reader.read_bytes 2
-    packet_id = binary.BIG_ENDIAN.uint16 data 0
+  constructor.deserialize_ reader/reader.BufferedReader:
+    packet_id = Packet.decode_uint16 reader
     super TYPE
 
   variable_header -> ByteArray:
-    data := ByteArray 2
-    binary.BIG_ENDIAN.put_uint16 data 0 packet_id
-    return data
+    return Packet.encode_uint16 packet_id
 
   payload -> ByteArray: return #[]
 
 class SubscribePacket extends Packet:
   static TYPE ::= 8
 
-  topic_filters/List/*<TopicFilter>*/
+  topics/List/*<TopicFilter>*/
   packet_id/int
 
-  constructor .topic_filters --.packet_id:
+  constructor .topics --.packet_id:
+    super TYPE --flags=0b0010
+
+  constructor.deserialize_ reader/reader.BufferedReader size/int:
+    packet_id = Packet.decode_uint16 reader
+    size -= 2
```

Comment:
I think this would be better written with a cursor that moves up towards size.

Reply:
done.